[Badge][material-next] Add Badge component

DiegoAndai opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

**Badge issue**: #37835 
**Material You umbrella issue**: #29345

### Add Badge component
- Copy Badge from `material`
- Migrate to Typescript
- Remove deprecated `components` and `componentsProps`
- Drop ThemeProvider in favor of CssVarsProvider

mnajdova commented on packages/mui-material-next/migration.md:
```diff
@@ -6,6 +6,58 @@ This is a reference guide on how to migrate from Material UI v5 to v6.
+
+Color options were changed to match Material You:
+
+- `"default"`, `"info"`, `"success"`, and `"warning"` were removed and are no longer supported out-of-the-box
```
Should we still keep these? We dsicussed that it may be useful to define these in the team. cc @danilo-leal

DiegoAndai replied:
I discussed with @danilo-leal and we will work on a Material UI Material You theme that makes upgrading smoother (for example, having these colors) and that makes sense with the Material You color system.

I'll be working on developing this, but there are still things to figure out, like for example which would be the default theme, the Material You default theme, or the Material UI MY theme.

I don't want to block the components progress while we figure out the Material UI MY theme, so for now I think we should continue the work using the Material You default theme (which doesn't have these colors) and when we have more clarity we can go back to this. What do you think?

mnajdova replied:
My worry with this is that we would introduce a breaking change and later add back these values. It could be frustrating for people to have to do some changes, and revert them later.

DiegoAndai replied:
You're right, what we should do then is add the `"info"`, `"success"`, and `"warning"` to the palette following Material You's custom colors guidelines: https://m3.material.io/styles/color/the-color-system/custom-colors.

Regarding the `"default"` color, that's not really a color but rather the badge has no background:

<img width="59" alt="Screenshot 2023-07-21 at 08 44 32" src="https://github.com/mui/material-ui/assets/16889233/12790971-a9d5-4950-872e-1d4dc39e30d7">

Do you think we should keep that one as well?

I'll work with @zanivan on this.

mnajdova replied:
In v5 we started with removing "default" from the color values, I wonder why it is still there in the `Badge`. The reasoning for it was that `default` is not a color, we should use as default some valid color value. I propose we drop it.

mnajdova commented on packages/mui-material-next/migration.md:
```diff
@@ -6,6 +6,58 @@ This is a reference guide on how to migrate from Material UI v5 to v6.
+
+Variants were renamed to match Material You:
+
+- `standard` renamed to `large`, which is still the default
+- `dot` renamed to `small`
```
We use `size` as a name for the sizes. I propose keeping all variant values as deprecated to easy up the migration, and use a new `size` prop for this instead. We can deprecate the whole `variant` prop with an appropriate deprecation message.

DiegoAndai replied:
@mnajdova I agree with naming the prop `size`.

About keeping the `variant` prop, I think it would be better to remove it right away, because:

- Migrating can be covered with a codemod
- Keeping it would complicate and bloat the code 

What do you think?

mnajdova replied:
I feel like we can likely get more people to try it early if we don't introduce many breaking changes with the first version, but happy to experiment with what you suggest.

I see two ways going forward:

- deprecate props and move forward with minimal breaking changes
- introduce breaking changes together with a codemod for them at the moment they are implemented

So we should either, add a codemod now, or deprication, one of these two needs to be implemented.

mnajdova reviewed (commented):
Looks great, left few comments for consideration.

DiegoAndai commented:
Hey @mnajdova I rolled back both deprecations 😊:

- Bring back the `variant` prop with a deprecation message
- Bring back `"info"`, `"warning"`, `"success"`, and `"default"` color options. I did maintain the change that `"error"` is the new default though as that's the only color used in the Material You specs.

You can check the new colors in the playground: https://deploy-preview-37850--material-ui.netlify.app/material-ui/react-badge/#material-you-version

@zanivan the colors are looking great! 🎉  here's a picture of them being used in the badges (in order: `"primary"`, `"error"`, `"info"`, `"warning"`, and `"success"`):

![badges](https://github.com/mui/material-ui/assets/16889233/a8116b8c-426b-4699-97a5-151edec2d5a0)

A couple of questions regarding the colors:

- May we use the tone 20 for the `on` dark tokens? (`onInfo`, `onWarning`, `onSuccess`). In Figma, the tone 15 is used, but all the other colors in Material You use tone 20. The picture is using the tone 20 of each color, I think it looks nice 👍🏼 
- The warning color on light mode is the only one that I think might need a little improving, I think it looks a little too brown and it should look more yellow or orange

zanivan commented:
> May we use the tone 20 for the on dark tokens?

Yep, absolutely! I'll update it on Figma.


> The warning color on light mode is the only one that I think might need a little improving, I think it looks a little too brown and it should look more yellow or orange

I'll change the Hue a bit—I wouldn't saturate it, though. MY seems less saturate, so yellow or orange could be tricky and look a bit off from the others, since those are colors with more perceived light.

DiegoAndai commented:
@mnajdova another question, regarding [this comment](https://github.com/mui/material-ui/pull/37850/files#diff-8638f7978a90833820b7b9e82d50a8128a29a08b6bff6b0ec10db8ea3cae5ae8R83), should we just remove the classes or is there anything else to do?

mnajdova commented on packages/mui-material-next/src/Badge/badgeClasses.ts:
```diff
@@ -0,0 +1,94 @@
+  'colorTertiary',
+  'overlapRectangular',
+  'overlapCircular',
+  // TODO: v6 remove the overlap value from these class keys
```
@DiegoAndai we don't want to have in v6 combination of props in the classes. If someone wants to combine these two, they can create selector containing both classes. Each of the keys in `classes` should define one prop + value combination.

zanivan commented:
@DiegoAndai I tried to bump the warning colors a bit to the yellow tones, I believe it works. Here's the [Figma link](https://www.figma.com/file/nNAatyLlRl3Tq4nGHgAmNh/Primitives?type=design&node-id=7031-45315&mode=design&t=ODSH1lKQordnLwMS-4)

![image](https://github.com/mui/material-ui/assets/37222944/b770c930-5c05-491f-9115-fc0cb8c3e766)

DiegoAndai commented:
@mnajdova I removed the combined classes, added the corresponding migration section, and updated tests. Here's a sandbox of the new usage creating a CSS selector to combine `anchorOrigin` and `overlap` classes: https://codesandbox.io/s/next-badge-overlap-anchor-class-k7vwf7?file=/src/App.tsx

@zanivan updated the warning color 🙌🏼:

![badges-2](https://github.com/mui/material-ui/assets/16889233/8c38d3ce-175c-4757-b97a-ff2bfc4bde52)

mnajdova commented on packages/mui-material-next/src/Badge/Badge.spec.tsx:
```diff
@@ -0,0 +1,73 @@
+<Badge color="primary" />;
+<Badge color="secondary" />;
+<Badge color="tertiary" />;
+<Badge color="error" />;
```
We can add the other colors too.

mnajdova reviewed (approved):
Looks good, left few final comments, nothing big. Well done! The PR description is very detailed, love it!
